retrofit/src/main/java/retrofit2/HttpServiceMethod.java

```diff
@@ -32,8 +35,28 @@
    */
   static <ResponseT, ReturnT> HttpServiceMethod<ResponseT, ReturnT> parseAnnotations(
       Retrofit retrofit, Method method, RequestFactory requestFactory) {
-    CallAdapter<ResponseT, ReturnT> callAdapter = createCallAdapter(retrofit, method);
-    Type responseType = callAdapter.responseType();
+    CallAdapter<ResponseT, ReturnT> callAdapter = null;
```

Comment:
Just a wrinkle: asymmetry on how this is set up feels awkward. No suggestions on how to do it better.

Reply:
Yeah I'm willing to muck up the internals instead of inventing some elaborate abstraction that will only be used by coroutines. If we ever need to generalize we can, but I don't want to jump that gun.

Reply:
I've refined this a bit by using subtypes. This method is still gnarly because of all the shared logic, but the created classes are small and focused and don't involve crazy conditionals

Reply:
This code actually reminds me of Retrofit internals circa 2012-2013... it can be cleaned up it's just a matter of effort but since it's isolated here for now I'm not worried about it.